File: binding/src/libqtexttable.cpp
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QTextCharFormat>
#include <QTextCursor>
#include <QTextDocument>
#include <QTextFrame>
#include <QTextTable>
#include <QTextTableCell>
#include <QTextTableFormat>
#include <qtexttable.h>
#include "libqtexttable.h"
#include "libqtexttable.hxx"

QTextTableCell* QTextTableCell_new() {
	 return new QTextTableCell();
}

QTextTableCell* QTextTableCell_new2(const QTextTableCell* o) {
	 return new QTextTableCell(*o);
}

void QTextTableCell_OperatorAssign(QTextTableCell* self, const QTextTableCell* o) {
	self->operator=(*o);
}

void QTextTableCell_SetFormat(QTextTableCell* self, const QTextCharFormat* format) {
	self->setFormat(*format);
}

QTextCharFormat* QTextTableCell_Format(const QTextTableCell* self) {
	return new QTextCharFormat(self->format());
}

int QTextTableCell_Row(const QTextTableCell* self) {
	return self->row();
}

int QTextTableCell_Column(const QTextTableCell* self) {
	return self->column();
}

int QTextTableCell_RowSpan(const QTextTableCell* self) {
	return self->rowSpan();
}

int QTextTableCell_ColumnSpan(const QTextTableCell* self) {
	return self->columnSpan();
}

bool QTextTableCell_IsValid(const QTextTableCell* self) {
	return self->isValid();
}

QTextCursor* QTextTableCell_FirstCursorPosition(const QTextTableCell* self) {
	return new QTextCursor(self->firstCursorPosition());
}

QTextCursor* QTextTableCell_LastCursorPosition(const QTextTableCell* self) {
	return new QTextCursor(self->lastCursorPosition());
}

int QTextTableCell_FirstPosition(const QTextTableCell* self) {
	return self->firstPosition();
}

int QTextTableCell_LastPosition(const QTextTableCell* self) {
	return self->lastPosition();
}

bool QTextTableCell_OperatorEqual(const QTextTableCell* self, const QTextTableCell* other) {
	return self->operator==(*other);
}

bool QTextTableCell_OperatorNotEqual(const QTextTableCell* self, const QTextTableCell* other) {
	return self->operator!=(*other);
}

int QTextTableCell_TableCellFormatIndex(const QTextTableCell* self) {
	return self->tableCellFormatIndex();
}

void QTextTableCell_Delete(QTextTableCell* self) {
    delete self;
}

QTextTable* QTextTable_new(QTextDocument* doc) {
	 return new QTextTable(doc);
}

libqt_string QTextTable_Tr(const char* s) {
	QString _ret = QTextTable::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QTextTable_Resize(QTextTable* self, int rows, int cols) {
	self->resize(rows, cols);
}

void QTextTable_InsertRows(QTextTable* self, int pos, int num) {
	self->insertRows(pos, num);
}

void QTextTable_InsertColumns(QTextTable* self, int pos, int num) {
	self->insertColumns(pos, num);
}

void QTextTable_AppendRows(QTextTable* self, int count) {
	self->appendRows(count);
}

void QTextTable_AppendColumns(QTextTable* self, int count) {
	self->appendColumns(count);
}

void QTextTable_RemoveRows(QTextTable* self, int pos, int num) {
	self->removeRows(pos, num);
}

void QTextTable_RemoveColumns(QTextTable* self, int pos, int num) {
	self->removeColumns(pos, num);
}

void QTextTable_MergeCells(QTextTable* self, int row, int col, int numRows, int numCols) {
	self->mergeCells(row, col, numRows, numCols);
}

void QTextTable_MergeCells2(QTextTable* self, const QTextCursor* cursor) {
	self->mergeCells(*cursor);
}

void QTextTable_SplitCell(QTextTable* self, int row, int col, int numRows, int numCols) {
	self->splitCell(row, col, numRows, numCols);
}

int QTextTable_Rows(const QTextTable* self) {
	return self->rows();
}

int QTextTable_Columns(const QTextTable* self) {
	return self->columns();
}

QTextTableCell* QTextTable_CellAt(const QTextTable* self, int row, int col) {
	return new QTextTableCell(self->cellAt(row, col));
}

QTextTableCell* QTextTable_CellAt2(const QTextTable* self, int position) {
	return new QTextTableCell(self->cellAt(position));
}

QTextTableCell* QTextTable_CellAt3(const QTextTable* self, const QTextCursor* c) {
	return new QTextTableCell(self->cellAt(*c));
}

QTextCursor* QTextTable_RowStart(const QTextTable* self, const QTextCursor* c) {
	return new QTextCursor(self->rowStart(*c));
}

QTextCursor* QTextTable_RowEnd(const QTextTable* self, const QTextCursor* c) {
	return new QTextCursor(self->rowEnd(*c));
}

void QTextTable_SetFormat(QTextTable* self, const QTextTableFormat* format) {
	self->setFormat(*format);
}

QTextTableFormat* QTextTable_Format(const QTextTable* self) {
	return new QTextTableFormat(self->format());
}

libqt_string QTextTable_Tr2(const char* s, const char* c) {
	QString _ret = QTextTable::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QTextTable_Tr3(const char* s, const char* c, int n) {
	QString _ret = QTextTable::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QTextTable_Delete(QTextTable* self) {
    delete self;
}

File: binding/src/libqtexttable.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qtexttablecell.html
pub const qtexttablecell = struct {

    /// New constructs a new QTextTableCell object.
    pub fn New() QtC.QTextTableCell {
        return qtc.QTextTableCell_new();
    }


    /// New2 constructs a new QTextTableCell object.
    pub fn New2(o: ?*anyopaque) QtC.QTextTableCell {
        return qtc.QTextTableCell_new2(@ptrCast(o));
    }


    pub fn OperatorAssign(self: ?*anyopaque, o: ?*anyopaque) void {
        qtc.QTextTableCell_OperatorAssign(@ptrCast(self), @ptrCast(o));
    }

    pub fn SetFormat(self: ?*anyopaque, format: ?*anyopaque) void {
        qtc.QTextTableCell_SetFormat(@ptrCast(self), @ptrCast(format));
    }

    pub fn Format(self: ?*anyopaque, ) QtC.QTextCharFormat {
        return qtc.QTextTableCell_Format(@ptrCast(self));
    }

    pub fn Row(self: ?*anyopaque, ) i32 {
        return qtc.QTextTableCell_Row(@ptrCast(self));
    }

    pub fn Column(self: ?*anyopaque, ) i32 {
        return qtc.QTextTableCell_Column(@ptrCast(self));
    }

    pub fn RowSpan(self: ?*anyopaque, ) i32 {
        return qtc.QTextTableCell_RowSpan(@ptrCast(self));
    }

    pub fn ColumnSpan(self: ?*anyopaque, ) i32 {
        return qtc.QTextTableCell_ColumnSpan(@ptrCast(self));
    }

    pub fn IsValid(self: ?*anyopaque, ) bool {
        return qtc.QTextTableCell_IsValid(@ptrCast(self));
    }

    pub fn FirstCursorPosition(self: ?*anyopaque, ) QtC.QTextCursor {
        return qtc.QTextTableCell_FirstCursorPosition(@ptrCast(self));
    }

    pub fn LastCursorPosition(self: ?*anyopaque, ) QtC.QTextCursor {
        return qtc.QTextTableCell_LastCursorPosition(@ptrCast(self));
    }

    pub fn FirstPosition(self: ?*anyopaque, ) i32 {
        return qtc.QTextTableCell_FirstPosition(@ptrCast(self));
    }

    pub fn LastPosition(self: ?*anyopaque, ) i32 {
        return qtc.QTextTableCell_LastPosition(@ptrCast(self));
    }

    pub fn OperatorEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QTextTableCell_OperatorEqual(@ptrCast(self), @ptrCast(other));
    }

    pub fn OperatorNotEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QTextTableCell_OperatorNotEqual(@ptrCast(self), @ptrCast(other));
    }

    pub fn Begin(self: ?*anyopaque, ) QtC.QTextFrame__iterator {
        return qtc.QTextTableCell_Begin(@ptrCast(self));
    }

    pub fn End(self: ?*anyopaque, ) QtC.QTextFrame__iterator {
        return qtc.QTextTableCell_End(@ptrCast(self));
    }

    pub fn TableCellFormatIndex(self: ?*anyopaque, ) i32 {
        return qtc.QTextTableCell_TableCellFormatIndex(@ptrCast(self));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QTextTableCell_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qtexttable.html
pub const qtexttable = struct {

    /// New constructs a new QTextTable object.
    pub fn New(doc: ?*anyopaque) QtC.QTextTable {
        return qtc.QTextTable_new(@ptrCast(doc));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QTextTable_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qtexttable.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Resize(self: ?*anyopaque, rows: i32, cols: i32) void {
        qtc.QTextTable_Resize(@ptrCast(self), rows, cols);
    }

    pub fn InsertRows(self: ?*anyopaque, pos: i32, num: i32) void {
        qtc.QTextTable_InsertRows(@ptrCast(self), pos, num);
    }

    pub fn InsertColumns(self: ?*anyopaque, pos: i32, num: i32) void {
        qtc.QTextTable_InsertColumns(@ptrCast(self), pos, num);
    }

    pub fn AppendRows(self: ?*anyopaque, count: i32) void {
        qtc.QTextTable_AppendRows(@ptrCast(self), count);
    }

    pub fn AppendColumns(self: ?*anyopaque, count: i32) void {
        qtc.QTextTable_AppendColumns(@ptrCast(self), count);
    }

    pub fn RemoveRows(self: ?*anyopaque, pos: i32, num: i32) void {
        qtc.QTextTable_RemoveRows(@ptrCast(self), pos, num);
    }

    pub fn RemoveColumns(self: ?*anyopaque, pos: i32, num: i32) void {
        qtc.QTextTable_RemoveColumns(@ptrCast(self), pos, num);
    }

    pub fn MergeCells(self: ?*anyopaque, row: i32, col: i32, numRows: i32, numCols: i32) void {
        qtc.QTextTable_MergeCells(@ptrCast(self), row, col, numRows, numCols);
    }

    pub fn MergeCells2(self: ?*anyopaque, cursor: ?*anyopaque) void {
        qtc.QTextTable_MergeCells2(@ptrCast(self), @ptrCast(cursor));
    }

    pub fn SplitCell(self: ?*anyopaque, row: i32, col: i32, numRows: i32, numCols: i32) void {
        qtc.QTextTable_SplitCell(@ptrCast(self), row, col, numRows, numCols);
    }

    pub fn Rows(self: ?*anyopaque, ) i32 {
        return qtc.QTextTable_Rows(@ptrCast(self));
    }

    pub fn Columns(self: ?*anyopaque, ) i32 {
        return qtc.QTextTable_Columns(@ptrCast(self));
    }

    pub fn CellAt(self: ?*anyopaque, row: i32, col: i32) QtC.QTextTableCell {
        return qtc.QTextTable_CellAt(@ptrCast(self), row, col);
    }

    pub fn CellAt2(self: ?*anyopaque, position: i32) QtC.QTextTableCell {
        return qtc.QTextTable_CellAt2(@ptrCast(self), position);
    }

    pub fn CellAt3(self: ?*anyopaque, c: ?*anyopaque) QtC.QTextTableCell {
        return qtc.QTextTable_CellAt3(@ptrCast(self), @ptrCast(c));
    }

    pub fn RowStart(self: ?*anyopaque, c: ?*anyopaque) QtC.QTextCursor {
        return qtc.QTextTable_RowStart(@ptrCast(self), @ptrCast(c));
    }

    pub fn RowEnd(self: ?*anyopaque, c: ?*anyopaque) QtC.QTextCursor {
        return qtc.QTextTable_RowEnd(@ptrCast(self), @ptrCast(c));
    }

    pub fn SetFormat(self: ?*anyopaque, format: ?*anyopaque) void {
        qtc.QTextTable_SetFormat(@ptrCast(self), @ptrCast(format));
    }

    pub fn Format(self: ?*anyopaque, ) QtC.QTextTableFormat {
        return qtc.QTextTable_Format(@ptrCast(self));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QTextTable_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qtexttable.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QTextTable_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qtexttable.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QTextTable_Delete(@ptrCast(self));
    }
};
